### PR TITLE
Improve few diagnostics

### LINKED ISSUE #48286: Diagnostic improvements for records
Suggestions for diagnostic improvements.

1. [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8kAmIA1AD6YB28CFEANgAQJwDGA9ggQwMIMDeAXwCwAKAACAZiasOXACIMQPfsJGigA=)

    **code:**

    ```csharp
    internal record C {}
    public record D : C {}
    ```

    **suggested improvements:**

    1. Change "class" to "type" in `base class 'C' is less accessible than class 'D'`
    2. Don't leak the unspeakable name `<Clone>$` in the message.

2. This is an internal improvement only (i.e. not visible nor changes any behavior). Rename `ERR_DuplicateNameInClass` to `ERR_DuplicateNameInType`. The error message already mentions "type" instead of "class".

3. [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8kAmIA1AD4ACAzAAQJwDGA9ggdQELRwCMA3ALAAoKrQbNWHKHABM/AYOEAHCAhiYIAGxFMW1AMoBXYNRDtOveTSUq1muttYGjJidNlA=)

    **code:**

    ```csharp
    public record Base1;
    public record Base2;

    public partial record Sub : Base1;
    public partial record Sub : Base2;
    ```

    **suggested improvements:**

    1. Change "classes" to "types" in `Partial declarations of '{0}' must not specify different base classes`

4. [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8kAmIA1AD6YB28C2EAxnAAQCSjA3owL4CwAUAnDoB7BAUYBBdlz58ADhAQxMEADaMAAgCZGAEQA8AFQB8jAO4ALREwOMQLJBKmcgA===)

    **code:**

    ```csharp
    interface I { }
    record A { }

    partial class D<T> where T : I, A { }
    ```

    **suggested improvements:**

    1. Change "class type constraint" to "type constraint" in `The class type constraint 'A' must come before any other constraints`.

### PR DESCRIPTION
**Pull request by Youssef1313**
I did part of the changes I suggested in #48286. The others in the issue aren't really important I think. But let me know if you think otherwise.

Fixes #48286

### CODE REVIEW

**Comment by Youssef1313** [comment]
Closing and re-opening for a new build. `(D:\workspace\_work\1\s\artifacts\toolset\restore.proj : error : Failed to retrieve information about 'Microsoft.DotNet.Arcade.Sdk' from remote source 'https://pkgs.dev.azure.com/azure-public/3ccf6661-f8ce-4e8a-bb2e-eff943ddd3c7/_packaging/58ca65bb-e6c1-4210-88ac-fa55c1cd7877/nuget/v3/flat2/microsoft.dotnet.arcade.sdk/index.json'.)`

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

**Comment by 333fred** [comment]
Would you mind going through the test cases that use these diagnostics and updating the messages?

**Comment by Youssef1313** [comment]
> Would you mind going through the test cases that use these diagnostics and updating the messages?

Sure. Let me know if you want to update the enum member names as well.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 1]
```diff
@@ -547,7 +547,7 @@
     <value>Inconsistent accessibility: parameter type '{1}' is less accessible than delegate '{0}'</value>
   </data>
   <data name="ERR_BadVisBaseClass" xml:space="preserve">
-    <value>Inconsistent accessibility: base class '{1}' is less accessible than class '{0}'</value>
+    <value>Inconsistent accessibility: base type '{1}' is less accessible than type '{0}'</value>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:550]: >type [](start = 44, length = 4)

I do not think we should make this change. A record class is a class.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 2]
```diff
@@ -1138,7 +1138,7 @@
     <value>Partial declarations of '{0}' have conflicting accessibility modifiers</value>
   </data>
   <data name="ERR_PartialMultipleBases" xml:space="preserve">
-    <value>Partial declarations of '{0}' must not specify different base classes</value>
+    <value>Partial declarations of '{0}' must not specify different base types</value>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:1141]: >types [](start = 73, length = 5)

I do not think we should make this change. A record class is a class and this error isn't about an interface in the base list, but specifically about the class type.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 3]
```diff
@@ -1276,7 +1276,7 @@
     <value>Duplicate constraint '{0}' for type parameter '{1}'</value>
   </data>
   <data name="ERR_ClassBoundNotFirst" xml:space="preserve">
-    <value>The class type constraint '{0}' must come before any other constraints</value>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:1279]: >class  [](start = 15, length = 6)

I do not think we should make his change. This error is specifically about class types (record classes are classes), not about interface types, for example.

**Review by AlekseyTs — APPROVED** [review]
I do not think the proposed changes are actually an improvement. In my opinion the existing wording is precise and correct.

**Comment by AlekseyTs** [comment]
> This makes diagnostics consistent with the changes already done in #46341

Didn't go through the changes in it, so it is quite possible that some or even all changes there make sense, or made sense at the time. I don't think there is a goal to get to "consistency" where every mention of a "class" is replaced with "type". Especially that now we have record classes and record structures, I don't believe there is a good reason to avoid referring to record classes as classes in cases when "record-ness" of the type isn't really important, by comparison to other aspects of the type.

**Comment by Youssef1313** [comment]
Got your point. But this probably means *some* of the changes in #46341 shouldn't have been made?

**Comment by AlekseyTs** [comment]
> Got your point. But this probably means some of the changes in #46341 shouldn't have been made?

Quite possible, however, it looks like some of the messages modified in this PR were looked at in context of that PR. They were not touched at the end. There are commits referring to ERR_BadVisBaseClass and ERR_ClassBoundNotFirst, and ERR_PartialMultipleBases was left unchanged based on my feedback (https://github.com/dotnet/roslyn/pull/46341#discussion_r460996780).

**Comment by Youssef1313** [comment]
Thanks for the additional info @AlekseyTs! That makes much sense.

**Pull request closed without merging**